cpp/include/cudf/join/sort_merge_join.hpp

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/export.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+#include <optional>
+#include <variant>
+
+namespace CUDF_EXPORT cudf {
+
+/**
+ * @addtogroup column_join
+ * @{
+ * @file
+ */
+
+/**
+ * @brief Class that implements sort-merge algorithm for table joins
+ */
+class sort_merge_join {
+ public:
+  /**
+   * @brief Construct a sort-merge join object that pre-processes the right table
+   * on creation, and can be used on subsequent join operations with multiple
+   * left tables.
+   *
+   * @note The `sort_merge_join` object must not outlive the table viewed by `right`,
+   * else behavior is undefined.
+   *
+   * @param right The right table
+   * @param is_right_sorted Enum to indicate if right table is pre-sorted
+   * @param compare_nulls Controls whether null join-key values should match or not
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   */
+  sort_merge_join(table_view const& right,
+                  sorted is_right_sorted,
+                  null_equality compare_nulls  = null_equality::EQUAL,
+                  rmm::cuda_stream_view stream = cudf::get_default_stream());
+
+  /**
+   * Returns the row indices that can be used to construct the result of performing
```

Comment:
Not sure if this will trigger error in CI doxygen check?
```suggestion
   * @brief Returns the row indices that can be used to construct the result of performing
```